plugins/UsersManager/API.php

```diff
@@ -727,28 +803,195 @@ public function getUserLoginFromUserEmail($userEmail)
+
+        $this->model->deleteUserAccess($userLogin, $idSites);
+
+        if ($access === 'noaccess') {
+            // if the access is noaccess then we don't save it as this is the default value
+            // when no access are specified
+            Piwik::postEvent('UsersManager.removeSiteAccess', array($userLogin, $idSites));
+        } else {
+            $role = array_shift($roles);
+            $this->model->addUserAccess($userLogin, $role, $idSites);
+        }
+
+        if (!empty($capabilities)) {
+            $this->addCapabilities($userLogin, $capabilities, $idSites);
+        }
+
+        // we reload the access list which doesn't yet take in consideration this new user access
+        $this->reloadPermissions();
+    }
+
+    /**
+     * Adds the given capabilities to the given user for the given sites.
+     * The capability will be added only when the user also has access to a site, for example View, Write, or Admin.
+     * Note: You can neither add any capability to a super user, nor to the anonymous user.
+     * Note: If the user has assigned a role which already grants the given capability, the capability will not be added in
+     * the backend.
+     *
+     * @param string $userLogin The user login
+     * @param string|string[] $capabilities  To fetch a list of available capabilities call "UsersManager.getAvailableCapabilities".
+     * @param int|int[] $idSites
+     * @throws Exception
+     */
+    public function addCapabilities($userLogin, $capabilities, $idSites)
+    {
+        $idSites = $this->getIdSitesCheckAdminAccess($idSites);
+
+        if ($userLogin == 'anonymous') {
+            throw new Exception(Piwik::translate("UsersManager_ExceptionAnonymousNoCapabilities"));
+        }
+
         $this->checkUserExists($userLogin);
         $this->checkUserHasNotSuperUserAccess($userLogin);
 
-        if ($userLogin == 'anonymous'
-            && $access == 'admin'
-        ) {
-            throw new Exception(Piwik::translate("UsersManager_ExceptionAdminAnonymous"));
+        if (!is_array($capabilities)){
+            $capabilities = array($capabilities);
+        }
+
+        foreach ($capabilities as $entry) {
+            $this->capabilityProvider->checkValidCapability($entry);
+        }
+
+        $sites = $this->model->getSitesAccessFromUser($userLogin);
+        $roleIds = $this->roleProvider->getAllRoleIds();
+
+        $sitesIdWithRole = array();
+        $sitesIdWithCapability = array();
+        foreach ($sites as $site) {
+            if (in_array($site['access'], $roleIds, true)) {
+                $sitesIdWithRole[(int) $site['site']] = $site['access'];
+            } else {
+                if (!isset($sitesIdWithCapability[(int) $site['site']])) {
+                    $sitesIdWithCapability[(int) $site['site']] = array();
+                }
+                $sitesIdWithCapability[(int) $site['site']][] = $site['access'];
+            }
         }
```

Comment:
I think this chunk of code could probably be refactored into a private method.

Reply:
I was thinking about that... but it wouldn't really be re-used anywhere and I would kind of need to split it into two private methods and go over the for loop in each method as they work on different variables (unless one private method returned an array of two variables). Didn't see much benefit of it then to put it into a private method but could do it.

Reply:
I figured you could have one method like:

```
private function getRolesAndCapabilitiesFor($userLogin)
{
        $sites = $this->model->getSitesAccessFromUser($userLogin);
        $roleIds = $this->roleProvider->getAllRoleIds();

        $sitesIdWithRole = array();
        $sitesIdWithCapability = array();
        foreach ($sites as $site) {
            if (in_array($site['access'], $roleIds, true)) {
                $sitesIdWithRole[(int) $site['site']] = $site['access'];
            } else {
                if (!isset($sitesIdWithCapability[(int) $site['site']])) {
                    $sitesIdWithCapability[(int) $site['site']] = array();
                }
                $sitesIdWithCapability[(int) $site['site']][] = $site['access'];
            }
         }
    return [$sitesIdWithRole, $sitesIdWithCapability];
}
```